lib/IDE/CodeCompletionResultType.cpp

```diff
@@ -11,50 +11,261 @@
+        USRTypes.push_back(USRBasedType::fromType(Superclass, Arena));
+      }
+      for (auto Proto : OpaqueType->getConformsTo()) {
+        USRTypes.push_back(
+            USRBasedType::fromType(Proto->getDeclaredInterfaceType(), Arena));
+      }
+      // Archetypes are also be used to model generic return types, in which
+      // case they don't have any conformsTo entries. We simply ignore those.
+      if (!USRTypes.empty()) {
+        ContextualTypes.emplace_back(USRTypes, /*IsConvertible=*/false);
+      }
+    }
+  }
+}
+
+TypeRelation
+USRBasedTypeContext::typeRelation(const USRBasedType *ResultType) const {
+  const USRBasedType *VoidType = Arena.getVoidType();
+  if (ResultType == VoidType) {
+    // Void is not convertible to anything and we don't report Void <-> Void
+    // identical matches (see USRBasedType::typeRelation). So we don't have to
+    // check anything if the result returns Void.
+    return TypeRelation::Unknown;
+  }
+
+  TypeRelation Res = TypeRelation::Unknown;
+  for (auto &ContextualType : ContextualTypes) {
+    Res = std::max(Res, ContextualType.typeRelation(ResultType, VoidType));
+  }
+  return Res;
+}
+
+// MARK: - USRBasedTypeArena
+
+USRBasedTypeArena::USRBasedTypeArena() {
+  // '$sytD' is the USR of the Void type.
+  VoidType = USRBasedType::fromUSR("$sytD", {}, *this);
+}
+
+const USRBasedType *USRBasedTypeArena::getVoidType() const { return VoidType; }
+
+// MARK: - USRBasedType
+
+const USRBasedType *USRBasedType::null(USRBasedTypeArena &Arena) {
+  return USRBasedType::fromUSR(/*USR=*/"", /*Supertypes=*/{}, Arena);
+}
+
+const USRBasedType *
+USRBasedType::fromUSR(StringRef USR, ArrayRef<const USRBasedType *> Supertypes,
+                      USRBasedTypeArena &Arena) {
+  auto ExistingTypeIt = Arena.CanonicalTypes.find(USR);
+  if (ExistingTypeIt != Arena.CanonicalTypes.end()) {
+    assert(ArrayRef<const USRBasedType *>(ExistingTypeIt->second->Supertypes) ==
+               Supertypes &&
+           "Same USR but different supertypes?");
+    return ExistingTypeIt->second;
+  }
+  // USR and Supertypes need to be allocated in the arena to be passed into the
+  // USRBasedType constructor. The elements of Supertypes are already allocated
+  // in the arena.
+  USR = USR.copy(Arena.Allocator);
+  Supertypes = Supertypes.copy(Arena.Allocator);
+
+  const USRBasedType *Result =
+      new (Arena.Allocator) USRBasedType(USR, Supertypes);
+  Arena.CanonicalTypes[USR] = Result;
+  return Result;
+}
+
+const USRBasedType *USRBasedType::fromType(Type Ty, USRBasedTypeArena &Arena) {
+  if (!Ty) {
+    return USRBasedType::null(Arena);
+  }
+
+  // USRBasedTypes are backed by canonical types so that equivalent types have
+  // the same USR.
+  Ty = Ty->getCanonicalType();
+
+  // For opaque types like 'some View', consider them equivalent to 'View'.
+  if (auto OpaqueType = Ty->getAs<ArchetypeType>()) {
+    if (auto Existential = OpaqueType->getExistentialType()) {
```

Comment:
Does this support protocol with associated types? Adding a test case would be great.

Reply:
It does, but while playing around with them I discovered another bug
```swift
public protocol ProtoWithAssocType {
  associatedtype MyAssoc
}


public func makeProtoWithAssocType() -> some ProtoWithAssocType { return StructWithAssocType() }

func protoWithAssocTypeInGenericContext<T: ProtoWithAssocType>() -> T {
   return #^COMPLETE^#
 }
```
would suggest `makeProtoWithAssocType` as “Identical” even though it is not valid here. The problem was that we were doing some of the special opaque return type handling for USRBasedTypes for all archetypes. Fix was to only apply them to `OpaqueTypeArchetypeType`. This regressed test cases which are using a generic context, but I think that’s fine since the main known limitation of the USR-based type comparisons is that they don’t support generic contexts. https://github.com/apple/swift/pull/41000/commits/5a1dd49d99acc3da122f20eb8fef81c3b44b94a6